### PR TITLE
Added functionality for data parallel table

### PR DESCRIPTION
**Pull request by csarofeen**
Added:
 Scatter of input across dims other than 0
 Arbitrary ordered input

Not supported: keyword arguments still not supported

For input:
Variables are scattered
Tensors are broadcasted
Primitive variables are broadcasted
All others are shallow copied
Supports nested lists/tuples and maintains structure

Gather
Nested lists, nested tuples, None and Variables are supported.

### CODE REVIEW

File: torch/cuda/comm.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ def scatter(tensor, devices, chunk_sizes=None, dim=0):
         assert min(chunk_sizes) > 0, "got a negative chunk_size"
         chunks = [tensor.narrow(dim, start - size, size)
                   for start, size in zip(_accumulate(chunk_sizes), chunk_sizes)]
+    chunks = tuple(chunk.contiguous() for chunk in chunks)
```
Comment (apaszke) [inline torch/cuda/comm.py]: Why is that needed? THC will make it contiguous anyway, so there's no need for that additional python loop.

File: torch/nn/parallel/data_parallel.py [hunk 2]
```diff
@@ -24,23 +31,27 @@ class DataParallel(Module):
...
             output_device = device_ids[0]
+        self.dim = dim
         self.module = module
         self.device_ids = device_ids
         self.output_device = output_device
         if len(self.device_ids) == 1:
             self.module.cuda(device_ids[0])
 
     def forward(self, *inputs):
+        #leave non-Variable/non-tuples in place
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: nit: add a space after the hash

File: torch/nn/parallel/scatter_gather.py [hunk 3]
```diff
@@ -1,25 +1,36 @@
...
     """
     def scatter_map(obj):
         if isinstance(obj, Variable):
-            return Scatter(target_gpus)(obj)
-        return tuple(zip(*map(scatter_map, obj)))
+            return Scatter(target_gpus, dim=dim)(obj)
+        if isinstance(obj, tuple) or isinstance(obj, list):
+            return type(obj)(zip(*map(scatter_map, obj)))
+        if torch.is_tensor(obj):
+            return broadcast(obj, target_gpus)
```
Comment (apaszke) [inline torch/nn/parallel/scatter_gather.py]: I wouldn't say that this behaviour is an obvious choice for tensors. In most places we're treating Variables and Tensors in the same way. What's the use case for this?

File: torch/nn/parallel/scatter_gather.py [hunk 4]
```diff
@@ -1,25 +1,36 @@
...
     """Gathers variables from different GPUs on a specified device
        (-1 means the CPU).
     """
     def gather_map(outputs):
         out = outputs[0]
         if isinstance(out, Variable):
-            return Gather(target_device)(*outputs)
-        return type(out)(map(gather_map, zip(*outputs)))
+            return Gather(target_device, dim=dim)(*outputs)
+        if isinstance(out, tuple) or isinstance(out, list):
```
Comment (apaszke) [inline torch/nn/parallel/scatter_gather.py]: I think this check is unnecessary. You can add `if out is None` earlier and leave the `type(out)` line as a fallback.

File: torch/nn/parallel/data_parallel.py [hunk 5]
```diff
@@ -24,23 +31,27 @@ class DataParallel(Module):
...
         def _to_cuda(obj):
             if isinstance(obj, Variable):
                 return obj.cuda()
-            return tuple((map(_to_cuda, obj)))
+            if isinstance(obj, tuple):
+                return tuple((map(_to_cuda, obj)))
+            return obj
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: I think a tensor might still need a call to `.cuda()` if `self.device_ids[0]` is different than its device (e.g. it's on CPU).

File: torch/cuda/comm.py (hunk 1 above)
Comment (csarofeen) [inline torch/cuda/comm.py]: Worried about transfer performance. Transferring peer to peer of contiguous should be faster so would like to do before.

Comment (apaszke) [inline torch/cuda/comm.py]: Right, nvm. I forgot that THC does that only for CPU tensors.

File: torch/nn/parallel/scatter_gather.py (hunk 3 above)
Comment (csarofeen) [inline torch/nn/parallel/scatter_gather.py]: Variables seem to be the only obvious thing to scatter. Also there's no obvious way to broadcast anything that would be contained in a tensor in data parallel. Not sure of a specific use case but could imagine an easy method to broadcast something instead of scattering could be useful.

File: torch/nn/parallel/scatter_gather.py (hunk 4 above)
Comment (csarofeen) [inline torch/nn/parallel/scatter_gather.py]: Maybe it should be left and an error thrown if it gets to end without finding the right type?

File: torch/nn/parallel/scatter_gather.py (hunk 3 above)
Comment (csarofeen) [inline torch/nn/parallel/scatter_gather.py]: I haven't done kwargs yet, but was thinking. A good compromise may be to scatter all variables and tensors that come in as normal positional arguments and broadcast all variables and tensors that come in as keyword arguments. May do the second as a separate PR, as it requires processing the dict of kwargs and each entry as the inputs are now.

**Comment by csarofeen** [comment]
@apaszke @colesbury included the changes we discussed on slack.

File: torch/nn/parallel/data_parallel.py [hunk 6]
```diff
@@ -36,48 +43,71 @@ class DataParallel(Module):
...
 
         if len(self.device_ids) == 1:
             with torch.cuda.device(self.device_ids[0]):
                 inputs_cuda = _to_cuda(inputs)
-            return self.module(*inputs_cuda)
+            if kwargs:
+                gpu_dict = {}
+                for key in kwargs.keys():
+                    gpu_dict[key] = _to_cuda(kwargs[key])
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: I think you should call that in the `with` block. Otherwise it might end up on a different GPU.

File: torch/nn/parallel/parallel_apply.py [hunk 7]
```diff
@@ -8,31 +8,44 @@
     import Queue as queue
 
 
-def parallel_apply(modules, inputs):
+def parallel_apply(modules, inputs, kwargs):
```
Comment (szagoruyko) [inline torch/nn/parallel/parallel_apply.py]: I think kwargs should have a default value

File: torch/nn/parallel/parallel_apply.py [hunk 8]
```diff
@@ -8,31 +8,44 @@
...
-                                args=(module, input, results, lock))
-               for module, input in zip(modules, inputs)]
+    if kwargs is None:
+        threads = [threading.Thread(target=_worker,
+                                    args=(module, input, kwargs, results, lock))
+                   for module, input in zip(modules, inputs)]
+    else:
+        threads = [threading.Thread(target=_worker,
+                                    args=(module, input, kwargs, results, lock))
+                   for module, input, kwargs in zip(modules, inputs, kwargs)]
```
Comment (szagoruyko) [inline torch/nn/parallel/parallel_apply.py]: it is possible to avoid this complication of kwargs by predefining `module` functions as `partial` functions, might be cleaner

Comment (csarofeen) [inline torch/nn/parallel/parallel_apply.py]: Got it cleaned up, thanks.

**Comment by csarofeen** [comment]
@apaszke @szagoruyko I think I hit all your comments, is there anything else you think needs to be done for this PR?

File: test/test_nn.py [hunk 9]
```diff
@@ -230,7 +232,8 @@ def fw_hook(inc, h_module, input, output):
             self.assertIsInstance(output, Variable)
             self.assertTrue(h_module is module)
             self.assertEqual(input[0].data, torch.ones(5, 5))
-            self.assertEqual(output.data, torch.Tensor(5, 5).fill_(1 / (1 + 1 / math.e)))
+            self.assertEqual(
```
Comment (apaszke) [inline test/test_nn.py]: Can you please leave the lines as they were? We accept up to 120 chars in line and they're not very long.

File: torch/nn/parallel/data_parallel.py [hunk 10]
```diff
@@ -36,48 +43,71 @@ class DataParallel(Module):
...
         replicas = self.replicate(self.module, self.device_ids)
         scattered = self.scatter(inputs, self.device_ids)
+
+        gpu_dicts = None
+        if kwargs:
+            scatter_kwargs = {}
+            for key in kwargs.keys():
+                scatter_kwargs[key] = self.scatter(
+                    _to_cuda(kwargs[key]), self.device_ids)
+            gpu_dicts = tuple([
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: No need to materialize the list if you're going to make a tuple out of it.

File: torch/nn/parallel/data_parallel.py [hunk 11]
```diff
@@ -36,48 +43,71 @@ class DataParallel(Module):
...
+                dict([(key, values[i])
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: `{key: values[i] for key, values in scatter_kwargs.items()}`. Keep it in one line and use `k, v` if necessary.

File: torch/nn/parallel/data_parallel.py [hunk 12]
```diff
@@ -36,48 +43,71 @@ class DataParallel(Module):
...
+    def parallel_apply(self, replicas, inputs, kwargs):
+        return parallel_apply(replicas, inputs, kwargs)
 
     def gather(self, outputs, output_device):
-        return gather(outputs, output_device)
+        return gather(outputs, output_device, dim=self.dim)
 
 
-def data_parallel(module, inputs, device_ids, output_device=None):
+def data_parallel(module, inputs, device_ids, module_kwargs=None, output_device=None, dim=0):
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: I think it's better to add `module_kwargs` at the end of argument list just in case someone has been using the `output_device` arg.

File: torch/nn/parallel/data_parallel.py [hunk 13]
```diff
@@ -96,13 +126,27 @@ def data_parallel(module, inputs, device_ids, output_device=None):
...
+
+    gpu_dicts = None
+    if module_kwargs:
+        scatter_kwargs = {}
+        for key in module_kwargs.keys():
+            scatter_kwargs[key] = scatter(module_kwargs[key], device_ids, dim)
+        gpu_dicts = tuple([
+            dict([(key, values[i]) for key, values in scatter_kwargs.items()])
+            for i in device_ids
+        ])
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: See above.

File: torch/nn/parallel/parallel_apply.py [hunk 14]
```diff
@@ -8,31 +8,37 @@
...
+def parallel_apply(modules, inputs, kwargs_tup=None):
     assert len(modules) == len(inputs)
+    if kwargs_tup:
+        assert len(modules) == len(kwargs_tup)
+    else:
+        kwargs_tup = tuple({} for gpu in modules)
     # Fast track
     if len(modules) == 1:
-        return (modules[0](*inputs[0]),)
+        return (wrap(modules[0], *inputs[0], **kwargs_tup[0]), )
```
Comment (apaszke) [inline torch/nn/parallel/parallel_apply.py]: what is `wrap`?

File: torch/nn/parallel/parallel_apply.py [hunk 15]
```diff
@@ -8,31 +8,37 @@
     import Queue as queue
 
 
-def parallel_apply(modules, inputs):
+def parallel_apply(modules, inputs, kwargs_tup=None):
     assert len(modules) == len(inputs)
+    if kwargs_tup:
+        assert len(modules) == len(kwargs_tup)
+    else:
+        kwargs_tup = tuple({} for gpu in modules)
```
Comment (apaszke) [inline torch/nn/parallel/parallel_apply.py]: I think `({},) * len(modules)` is better.

File: torch/nn/parallel/parallel_apply.py [hunk 16]
```diff
@@ -8,31 +8,37 @@
...
     # Fast track
     if len(modules) == 1:
-        return (modules[0](*inputs[0]),)
+        return (wrap(modules[0], *inputs[0], **kwargs_tup[0]), )
 
     lock = threading.Lock()
     results = {}
 
-    def _worker(module, input, results, lock):
+    def _worker(module, input, results, lock, **kwargs):
```
Comment (apaszke) [inline torch/nn/parallel/parallel_apply.py]: I think it's better to pass the kwargs as a single dict argument (probably next to the input).

**Review by apaszke — COMMENTED** [review]
Can you please not change the formatting of the code that doesn't need to be touched by this PR? It makes the diffs smaller, and there's no point in doing that, because we're not merging any PRs that don't pass the linter checks.

**Comment by apaszke** [comment]
I think that once the comments are fixed and the conflict is resolved it's ready to merge.

File: torch/nn/parallel/scatter_gather.py [hunk 17]
```diff
@@ -1,25 +1,37 @@
...
 from ._functions import Scatter, Gather
+from torch.cuda.comm import broadcast
 
 
-def scatter(input, target_gpus):
-    """Slices a given variable into approximately equal chunks and distributes
-       them accross given GPUs
+def scatter(input, target_gpus, dim=0):
+    """
+    Slices all variables and tensors into approximately
```
Comment (csarofeen) [inline torch/nn/parallel/scatter_gather.py]: Thanks, fixed.

File: torch/nn/parallel/scatter_gather.py [hunk 18]
```diff
@@ -1,25 +1,37 @@
...
+      equal chunks and distributes them accross given GPUs.
+    Duplicates references to objects that are not variables
+      or tensors.
     """
     def scatter_map(obj):
         if isinstance(obj, Variable):
-            return Scatter(target_gpus)(obj)
-        return tuple(zip(*map(scatter_map, obj)))
+            return Scatter(target_gpus, dim=dim)(obj)
+        assert not torch.is_tensor(obj), "Tensors not supported in DataParallel"
```
Comment (csarofeen) [inline torch/nn/parallel/scatter_gather.py]: Thanks, fixed.

File: test/test_nn.py (hunk 9 above)
Comment (csarofeen) [inline test/test_nn.py]: Sorry about that, ran it though autopep and left the changes.

File: torch/nn/parallel/data_parallel.py (hunk 10 above)
Comment (csarofeen) [inline torch/nn/parallel/data_parallel.py]: Thanks, fixed.

File: torch/nn/parallel/data_parallel.py (hunk 11 above)
Comment (csarofeen) [inline torch/nn/parallel/data_parallel.py]: Thanks, fixed.

File: torch/nn/parallel/data_parallel.py (hunk 12 above)
Comment (csarofeen) [inline torch/nn/parallel/data_parallel.py]: Done.

File: torch/nn/parallel/data_parallel.py (hunk 13 above)
Comment (csarofeen) [inline torch/nn/parallel/data_parallel.py]: Done.

File: torch/nn/parallel/parallel_apply.py (hunk 15 above)
Comment (csarofeen) [inline torch/nn/parallel/parallel_apply.py]: Done.

File: torch/nn/parallel/parallel_apply.py (hunk 14 above)
Comment (csarofeen) [inline torch/nn/parallel/parallel_apply.py]: Remnant from the refactor, no idea how that got through and I didn't catch it on tests. Fixed.

File: torch/nn/parallel/parallel_apply.py (hunk 16 above)
Comment (csarofeen) [inline torch/nn/parallel/parallel_apply.py]: Done.

**Comment by apaszke** [comment]
Thanks!
